File: src/javascript/algorithms/array/47.removeDuplicatesSortedArray/removeDuplicatesSortedArray.js
// Two pointers TC: O(n) SC: O(1)
function removeDuplicates1(nums) {
    let left = 0;

    for (let j = 1; j < nums.length; j++) {
        if (nums[j] !== nums[left]) {
            left++;
            nums[left] = nums[j];
        }
    }

    return left + 1;
}

// Two pointers (previous element comparison) TC: O(n) SC: O(1)
function removeDuplicates2(nums) {
    let left = 1;

    for (let right = 1; right < nums.length; right++) {
        if (nums[right] !== nums[right - 1]) {
            nums[left++] = nums[right];
        }
    }

    return left;
}

const testCases = [
    { nums: [1, 1, 2],                     expectedK: 2, expectedArr: [1, 2] },
    { nums: [0, 0, 1, 1, 1, 2, 2, 3, 3, 4], expectedK: 5, expectedArr: [0, 1, 2, 3, 4] },
    { nums: [1, 2, 3, 4, 5],               expectedK: 5, expectedArr: [1, 2, 3, 4, 5] },
    { nums: [1],                            expectedK: 1, expectedArr: [1] },
    { nums: [7, 7, 7, 7],                  expectedK: 1, expectedArr: [7] },
];

function runTests(fn, label) {
    console.log(`--- ${label} ---`);
    for (const { nums, expectedK, expectedArr } of testCases) {
        const arr = [...nums];
        const k = fn(arr);
        const resultArr = arr.slice(0, k);
        const pass = k === expectedK && resultArr.every((v, i) => v === expectedArr[i]);
        console.log(`[${pass ? "PASS" : "FAIL"}] k=${k} array=[${resultArr}]`);
    }
}

runTests(removeDuplicates1, "removeDuplicates1 (left pointer)");
runTests(removeDuplicates2, "removeDuplicates2 (prev element comparison)");

File: src/python/algorithms/array/removeDuplicatesSortedArray/remove_duplicates_sorted_array.py
from typing import List


def remove_duplicates1(nums: List[int]) -> int:
    """Two pointers TC: O(n) SC: O(1)"""
    left = 0
    for j in range(1, len(nums)):
        if nums[j] != nums[left]:
            left += 1
            nums[left] = nums[j]
    return left + 1


def remove_duplicates2(nums: List[int]) -> int:
    """Two pointers (previous element comparison) TC: O(n) SC: O(1)"""
    left = 1
    for right in range(1, len(nums)):
        if nums[right] != nums[right - 1]:
            nums[left] = nums[right]
            left += 1
    return left


if __name__ == "__main__":
    test_cases = [
        ([1, 1, 2],                     2, [1, 2]),
        ([0, 0, 1, 1, 1, 2, 2, 3, 3, 4], 5, [0, 1, 2, 3, 4]),
        ([1, 2, 3, 4, 5],               5, [1, 2, 3, 4, 5]),
        ([1],                            1, [1]),
        ([7, 7, 7, 7],                  1, [7]),
    ]

    print("--- remove_duplicates1 (left pointer) ---")
    for nums, expected_k, expected_arr in test_cases:
        arr = nums[:]
        k = remove_duplicates1(arr)
        result_arr = arr[:k]
        status = "PASS" if k == expected_k and result_arr == expected_arr else "FAIL"
        print(f"[{status}] k={k} array={result_arr}")

    print("--- remove_duplicates2 (prev element comparison) ---")
    for nums, expected_k, expected_arr in test_cases:
        arr = nums[:]
        k = remove_duplicates2(arr)
        result_arr = arr[:k]
        status = "PASS" if k == expected_k and result_arr == expected_arr else "FAIL"
        print(f"[{status}] k={k} array={result_arr}")
